From c5e9a50802e2b95f657dfaa2eaba98378b0a845d Mon Sep 17 00:00:00 2001
From: GitLab <example@example.com>
Date: Sun, 26 Jul 2015 01:12:12 +0200
Subject: [PATCH 1/2] Use grack for authn/authz only

Actual Git content will be served by gitlab-git-http-server.
---
 lib/gitlab/backend/grack_auth.rb |    3 ++-
 lib/gitlab/backend/shell_env.rb  |    6 +++++-
 2 files changed, 7 insertions(+), 2 deletions(-)

diff --git a/lib/gitlab/backend/grack_auth.rb b/lib/gitlab/backend/grack_auth.rb
index 03cef30..92e967d 100644
--- a/lib/gitlab/backend/grack_auth.rb
+++ b/lib/gitlab/backend/grack_auth.rb
@@ -26,7 +26,8 @@ module Grack
       auth!
 
       if project && authorized_request?
-        @app.call(env)
+        # Tell gitlab-git-http-server the request is OK, and what the GL_ID is
+        return [200, { "Content-Type" => "text/plain" }, [Gitlab::ShellEnv.gl_id(@user)]]
       elsif @user.nil? && !@gitlab_ci
         unauthorized
       else
diff --git a/lib/gitlab/backend/shell_env.rb b/lib/gitlab/backend/shell_env.rb
index 17ec029..009a3ec 100644
--- a/lib/gitlab/backend/shell_env.rb
+++ b/lib/gitlab/backend/shell_env.rb
@@ -7,7 +7,7 @@ module Gitlab
     def set_env(user)
       # Set GL_ID env variable
       if user
-        ENV['GL_ID'] = "user-#{user.id}"
+        ENV['GL_ID'] = gl_id(user)
       end
     end
 
@@ -15,5 +15,9 @@ module Gitlab
       # Reset GL_ID env variable
       ENV['GL_ID'] = nil
     end
+
+    def gl_id(user)
+      "user-#{user.id}"
+    end
   end
 end
-- 
1.7.9.5

